### ISSUE TITLE
Bidirectional A* path reconstruction is incorrect

### ISSUE THREAD

**Issue by lorenzhs** [qiao/PathFinding.js#28]
When a node has been settled by both the forward and backward search, the search can be terminated, as the shortest path has been found. This does not mean, however, that this node must lie on the shortest path.

Take this example in a graph: http://www.cosketch.com/SavedImages/phS9Zraa.png
Forward search starts at s, inserts v into queue with weight 6 and t with weight 10. Backward search starts at t, inserts v into queue with weight 6 and s with weight 10. Forward search removes v from queue, finds t with weight 12, does not add to queue, settles v. Backward search removes v from queue, finds s with weight 12, does not add to queue, settles v. The search is aborted here, yet v is not on the shortest path.

Screenshot to illustrate the problem: http://i.imgur.com/N9FaLY6.png
The shortest path is obviously incorrectly drawn.

**Comment by surrim** [comment]
qiao wrote:
> Well it does find the path, it just shows the wrong one ;)

lorenzhs wrote:
> You are correct. The Bi-Directional A* Search is not guaranteed to find the shortest path and it has been noted in the README.

You implementation is not guaranteed to find the right path, because it has a bug ;)
![bug](http://www.image-load.net/users/public/images/ARxCiID8Sx.png)
The distance in this example is 14.24 = 10 + 3 * sqrt(2). The best should be 13.66 = 8 + 4 * sqrt(2).

The error is located here: https://github.com/qiao/PathFinding.js/blob/master/src/finders/BiAStarFinder.js#L113
"f" is the compared key on the heap. You change it without updating the heap. So it can't be moved to the right place. That means when you call pop() you don't always get the best f-value. When you can't directly remove it from the heap, you can add an "invalid" boolean to the neighbour. When you find this element you have to add it again to the open list(s). The same problems exists for non-bidirectional searches. But here it's much harder to find an example on smaller maps.

**Comment by lorenzhs** [comment]
True, but that might be a separate issue. What I said was that the implementation correctly computes the shortest path, but retrieves it incorrectly. Thus, the result is wrong. @surrim just proved that it doesn't even compute the correct result, so there are two things that need to be fixed.
